### PR TITLE
Elasticsearch ^8.0 support

### PR DESCRIPTION
**Pull request by jayjfletcher**
This pull request contains breaking changes.

Elasticsearch 8.x introduced many breaking changes to the client. It seems to have been a complete overhaul.

8.x also introduced the ClientInterface so there is no longer a need for the following files:

- src/Infrastructure/Elastic/ElasticClientFactory.php - removed
- src/Infrastructure/Elastic/ElasticClientBuilder.php- removed

and associated tests:

- tests/Unit/ElasticClientBuilderTest.php - removed
- tests/Unit/ElasticClientFactoryTest.php - removed

Client can now be injected via ClientInterface so ElasticClientBuilder was replaced with Elastic's native ClientBuilder and documentation has been updated to reflect this.

### CODE REVIEW

**Comment by Jeroen-G** [comment]
Whoah, you have put in a good amount of work! Thanks, I have approved the CI actions to run and give you some feedback and I will review the code asap.

File: src/Application/Results.php [hunk 1]
```diff
@@ -5,14 +5,15 @@
...
 
 class Results implements Countable
 {
     private array $rawResults;
 
-    public function __construct(array $rawResults)
+    public function __construct(array|Elasticsearch $rawResults)
     {
-        $this->rawResults = $rawResults;
+        $this->rawResults = $rawResults instanceof Elasticsearch ? $rawResults->asArray() : $rawResults;
```
Comment (sporchia) [inline src/Application/Results.php:16]:
```suggestion
        $this->rawResults = is_array($rawResults) ? $rawResults : $rawResults->asArray();
```
this will probably make github action happier

**Comment by sporchia** [comment]
@Jeroen-G I have been working on updating this PR to get 8.x into the lib. A few things related to CI, looks like there will need to be an update to a fair bit more as the `Client` class is the correct class to use, since `ClientInterface` doesn't have all the correct methods and static analysis. the `Client` class is now marked as `final` so as ES suggests the tests will have to be updated to mock the underlying HTTP requests instead of their `Client` class. 

Long story short, would you rather I create a PR into this PR, or just have a fresh PR (with these initial changes in it and the new adjustments)?

**Comment by Jeroen-G** [comment]
Closed in favour of #246
